### PR TITLE
reorder forward declarations to get rid of C++14-only auto return types

### LINKED ISSUE #2718: Some forward declarations in execution_fwd.hpp aren't C++11-compatible
Hoi,

my C++11-only builds of LGD with HPX are failing because return types are given as auto without a resolution of that type, e.g. https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/execution_fwd.hpp#L76

I'm not sure how to resolve this as the forward declarations seem to be a hornets' nest. Ideas? Thanks!

### PR DESCRIPTION
**Pull request by gentryx**
This fixes C++11-only builds (basically anything that's using CUDA). Closes #2718

### CODE REVIEW

**Comment by hkaiser** [comment]
@gentryx This is roughly what we've had before. Unfortunately this fails compiling on some platforms (sorry, don't remember which) because of incomplete types. Could you please change the PR such that both versions of the code are there, using a pp constant to switch between the two (`HPX_HAVE_CXX11`? - I know this does not exist yet, it would need to be added here: https://github.com/STEllAR-GROUP/hpx/blob/master/cmake/HPX_DetectCppDialect.cmake)?

**Comment by gentryx** [comment]
@hkaiser Are you referring by "this" to the use of decltype instead of auto for the return types in the customization points? Yeah, I can change that to be configurable, no problem. :-)

**Comment by hkaiser** [comment]
@gentry: I meant the whole forward declaration. Also, the use of `auto` without an explicit return type is intentional. So I'd appreciate it if you left the existing code unchanged as an alternative to the code you would like to submit.

**Comment by gentryx** [comment]
@hkaiser I think I don't quite understand. What I did was basically two things:

1. I changed the C++14-style use of "auto" as a return type to be C++-11-compliant. I can change that to use the C++14-style if available.

2. I rearranged the order of a ton of class/function definitions to make the C++11-style work (because that's referring to functions and classes that were previously defined after the customization points. This should not affect any platform, right? If it's working for my stricter builds, it should work on other platforms just as well. Or am I missing something?

**Comment by gentryx** [comment]
Oh, and I forgot:

3. I removed the forward declaration for the customization points. Do you want me to re-add that to execution_fwd.hpp? I suppose I can do that as well, no problem.

**Comment by hkaiser** [comment]
@gentry: 

- In C++11 you can use just `auto` without explicit return type specification if the body of the function is just a single return statement.
- The reordering you propose is exactly what we started off with but we had to change it to use the forward declaration as other compilers were complaining about incomplete types in C++14 mode
- I'd like to ask you to (optionally) leave the code as it is right now and introduce your fix via a preprocessor constant which switches between the two implementations

Thanks!

**Comment by gentryx** [comment]
@hkaiser: I've added preprocessor guards so that the customization points will again use forward declarations and "auto" if C++14 is available, but will fall back to a C++11-compatible style if necessary.

I've successfully tested this build with GCC 4.9.4 with and without CUDA 8.0 for the C++11-mode and with GCC 6.3.0 for a C++14 build. Please let me know if you want me to try further compiler versions. :-)

File: cmake/tests/cxx14_auto.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+////////////////////////////////////////////////////////////////////////////////
+
+template<typename T>
+auto check_f(T& i)
+{
+    T j = 1 + j;
+    return j;
+}
```
Comment (hkaiser) [inline cmake/tests/cxx14_auto.cpp]: This is not C++11. For this, the body of the function has to consist of only the return statement.

**Comment by hkaiser** [comment]
@gentry: should those changes be applied to the execution information and time execution customization points as well: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/execution_information_fwd.hpp, https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/timed_execution_fwd.hpp)?

**Comment by hkaiser** [comment]
> should those changes be applied to the execution information and timed execution customization points as well: https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/execution_information_fwd.hpp and https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/executors/timed_execution_fwd.hpp?

@gentryx: to clarify: you have applied different methods of resolving the issue at hand for the base execution customization points compared to how you resolved it for the timed execution and informational customization points. Why?

Also, the original code was using the `auto` return type deduction for functions on purpose (without explicit trailing return type). This improves error messages as the `customization_point<>::operator()()` implementations are not SFINAE'd out in case of some underlying problems. You propose to unconditionally replace this with explicit return type deductions (`-> decltype(...)`) for the customization points related to the timed execution and informational facilities. I understand that we have to add explicit return type deduction for some compilers, but we should leave the original code in place for conforming ones.

You have also removed the headers `timed_execution_fwd.hpp` and execution_information_fwd.hpp` from the documentation without re-adding the corresponding headers (which now have the doxygen comments). This removes those customization points from the generated docs. A similar problem now occurs related to the doxygen comments for the main customization points. Those comments are now in a new header file which is not visible to the documentation tool chain, again removing them from the generated docs.

File: cmake/tests/cxx14_auto.cpp (hunk 1 above)
Comment (hkaiser) [inline cmake/tests/cxx14_auto.cpp]: Let me rephrase this. If the feature test is for a C++14 feature (which is fine), why do you use the result of this feature test to add a workaround for a missing C++11 capability of a particular compiler? Shouldn't we rather feature test for the particular missing capabilities und use this instead to decide whether to activate the workaround?

Comment (gentryx) [inline cmake/tests/cxx14_auto.cpp]: Ah, got it. You're right, this is apparently a bug in NVCC. However, CMake isn't using NVCC to test for compiler flags (but rather the default C++ compiler), so a test here for that specific defect likely wouldn't yield us much.

If you object to this test as it is in the PR then I think it would be fine to just tell CUDA users to manually set "-DWITH_CXX11_AUTO=false". In that case I'd remove the new C++14 auto test. Your thoughts?

**Comment by gentryx** [comment]
@hkaiser I added explicit return types (via decltype(...)) for customization_point because of NVCC's shortcomings WRT auto return types. This seems to be the only solution that works with NVCC (so apparently NVCC is not standard compliant). However, this is conditional on the CMake check for C++14 auto support, and if such support is found the code after the preprocessor is identical to the original code. As mentioned above I can change the conditional from "C++14 auto" to "C++11 auto".

If this approach is fine for you then I can follow the same pattern for timed_execution_fwd.hpp and execution_information_fwd.hpp. I had these headers removed as my builds (g++ 4.9.4 with CUDA 8.0 g++ 6.2.0 and clang 4.0) work just fine with moving those functions into the corresponding headers, but I hold no stake in this, I'm fine with doing the same as I did for execution.hpp/execution_fwd.hpp.

**Comment by gentryx** [comment]
OK, looks like CircleCI is fine with the PR (the error it's throwing seems related to the Perl install?). :-)

Please advise on how to procedd WRT timed_execution_fwd.hpp/execution_information_fwd.hpp. Same pattern as for execution_fwd.hpp?

**Comment by hkaiser** [comment]
> OK, looks like CircleCI is fine with the PR (the error it's throwing seems related to the Perl install?). :-)

No idea what's wrong there... I have never seen this before.

> Please advise on how to procedd WRT timed_execution_fwd.hpp/execution_information_fwd.hpp. Same pattern as for execution_fwd.hpp?

Yes, please.

**Comment by hkaiser** [comment]
@gentryx: are you still working on this?

**Comment by gentryx** [comment]
@hkaiser Please take another look (if CircleCI succeeds). I updated this PR to HPX' trunk and implemented it the way we discussed earlier.

Tested with g++ 4.9 and CUDA, g++ 6.3 and clang++ 4.0.

**Comment by gentryx** [comment]
@hkaiser Looks like this one went through successfully. Please take another look.

File: cmake/HPX_AddConfigTest.cmake [hunk 2]
```diff
@@ -208,7 +209,7 @@ macro(hpx_check_for_cxx11_auto)
   add_hpx_config_test(HPX_WITH_CXX11_AUTO
     SOURCE cmake/tests/cxx11_auto.cpp
     FILE ${ARGN}
-    CMAKECXXFEATURE cxx_auto_type)
+    CMAKECXXFEATURE cxx11_auto_type)
```
Comment (hkaiser) [inline cmake/HPX_AddConfigTest.cmake]: There is no C++ feature detection property in cmake called `cxx11_auto_type` (please see https://cmake.org/cmake/help/v3.9/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

File: cmake/HPX_AddConfigTest.cmake [hunk 3]
```diff
@@ -511,6 +512,14 @@ macro(hpx_check_for_cxx11_std_unordered_set)
 endmacro()
 
 ###############################################################################
+macro(hpx_check_for_cxx14_auto)
+  add_hpx_config_test(HPX_WITH_CXX14_AUTO
+    SOURCE cmake/tests/cxx14_auto.cpp
+    FILE ${ARGN}
+    CMAKECXXFEATURE cxx14_auto_type)
```
Comment (hkaiser) [inline cmake/HPX_AddConfigTest.cmake]: There is no C++ feature detection property in cmake called `cxx14_auto_type` (please see https://cmake.org/cmake/help/v3.9/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

File: cmake/tests/cxx14_auto.cpp (hunk 1 above)
Comment (hkaiser) [inline cmake/tests/cxx14_auto.cpp]: I'm not opposed to this feature test. I'm opposed to using the results of this test to detect a NVCC problem related to a C++11 feature. Couldn't we just force HPX_WITH_CXX11_AUTO=OFF whenever HPX is configured with CUDA/NVCC enabled?

File: hpx/exception_info.hpp [hunk 4]
```diff
@@ -73,10 +74,10 @@ namespace hpx
           : public exception_info_node_base
         {
         public:
-            template <typename ...Tags, typename ...Types>
+            template <typename ...ErrorInfo>
```
Comment (hkaiser) [inline hpx/exception_info.hpp]: What is the rationale of changing this file?

File: hpx/parallel/executors/execution_fwd.hpp [hunk 5]
```diff
@@ -67,6 +68,7 @@ namespace hpx { namespace parallel { namespace execution
             }
         };
 
+#ifdef HPX_HAVE_CXX14_AUTO
```
Comment (hkaiser) [inline hpx/parallel/executors/execution_fwd.hpp]: Please don't use a macro named `_CXX14_...` to guard code which is C++11 compliant.

File: hpx/exception_info.hpp (hunk 4 above)
Comment (gentryx) [inline hpx/exception_info.hpp]: NVCC couldn'T compile this.

File: cmake/tests/cxx14_auto.cpp (hunk 1 above)
Comment (gentryx) [inline cmake/tests/cxx14_auto.cpp]: At CMake time we don't know if the user is going to use nvcc for their code.

I can drop the additional (misleading) CXX14 CMake test and then users can manually disable CXX11 auto support. Would that be better?

File: hpx/parallel/executors/execution_fwd.hpp (hunk 5 above)
Comment (gentryx) [inline hpx/parallel/executors/execution_fwd.hpp]: ACK, will fix.

**Comment by gentryx** [comment]
I've changed the PR according to your comments, build failed due to the usual Perl install error. PTAL :-)

**Comment by gentryx** [comment]
@hkaiser Is there anything else you want changed for this PR or is this good to go?

**Comment by hkaiser** [comment]
@gentry: frankly I'm not too happy to relax the requirement for HPX to depend on `auto` in general. This is a problem specific to one compiler after all. But at this point I'm not sure how to proceed. Please give me a day or two to have a closer look and to play with this. 

The main problem is that we don't run the feature tests for the compiler used for compiling CUDA but for the base compiler NVCC is using to generate code...

**Comment by gentryx** [comment]
@hkaiser In this PR I'm using the CMake auto test only to switch between different definitions for the affected classes. Would you prefer if I introduced a second CMake test so that the current one is mandatory again and the second one can be overridden?

**Comment by hkaiser** [comment]
> Would you prefer if I introduced a second CMake test so that the current one is mandatory again and the second one can be overridden?

@gentryx , yah, that might be a good compromise.

File: hpx/parallel/executors/execution_information.hpp [hunk 6]
```diff
@@ -270,15 +323,99 @@ namespace hpx { namespace parallel { namespace execution
...
+        /// Set various modes of operation on the scheduler underneath the
+        /// given executor.
+        ///
+        /// \param exec     [in] The executor object to use.
+        /// \param mode     [in] The new mode for the scheduler to pick up
+        ///
+        /// \note This calls exec.set_scheduler_mode(mode) if it exists;
+        ///       otherwise it does nothing.
+        ///
+        constexpr detail::customization_point<
```
Comment (Naios) [inline hpx/parallel/executors/execution_information.hpp]: `HPX_CONSTEXPR`

Comment (gentryx) [inline hpx/parallel/executors/execution_information.hpp]: Yeah, maybe better do that in a separate PR instead of getting sidetracked here :-)

**Comment by gentryx** [comment]
@hkaiser I've reverted the change to cmake/HPX_PerformCxxFeatureTests.cmake and introduced a new CMake option HPX_WITH_CXX11_AUTO_RETURN_VALUE which is only used in the code I've changed. The option is on by default, but can be overridden by the user.

**Comment by gentryx** [comment]
Builds are successful (modulo the usual Perl woes).

**Comment by hkaiser** [comment]
@gentryx: thanks a lot, I'll have a look later today. I still don't understand where those Perl problems come from, though - this happens on this branch only.

**Comment by gentryx** [comment]
@hkaiser nag nag

**Comment by hkaiser** [comment]
@gentryx: sorry for the delay - I'll do it asap.

**Comment by gentryx** [comment]
@hkaiser I can understand that you're conservative about merging a failing build since you don't want to break the trunk. But I'd like to add that my HPX+LGD builds have been broken now for quite a while, or in other words: the HPX trunk is broken for CUDA since then.

My best guess is that for my build I see "CIRCLE_BUILD_IMAGE=ubuntu-14.04" while in https://circleci.com/gh/STEllAR-GROUP/hpx/7572?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link I see "CIRCLE_BUILD_IMAGE=ubuntu-12.04", so I'd say the build failures are unrelated to my PR and simply a sign that HPX builds with newer CircleCI images will fail.

**Comment by gentryx** [comment]
I've now switched my HPX builds to use Ubuntu 12.04 if that's required to get this merged. You should file a separate ticket to make sure HPX builds with 14.04 too because I assume that 12.04 will some day no longer be available.

**Comment by hkaiser** [comment]
@gentryx: sorry for going back and forth on this, but wouldn't it be sufficient to apply this patch to the main CMakeLists.txt in order to disable the functionality for NVCC only:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index fd416de2be..fcfefe5472 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1084,6 +1084,7 @@ if(HPX_WITH_CUDA)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};--default-stream per-thread)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-lcudadevrt)
     set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-rdc=true)
+    set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-DHPX_HAVE_CXX11_AUTO_RETURN_VALUE)
   else()
     hpx_add_target_compile_option(-DBOOST_THREAD_USES_MOVE)
     hpx_add_target_compile_option(--cuda-path=${CUDA_TOOLKIT_ROOT_DIR})
@@ -1112,13 +1113,6 @@ else()
   endif()
 endif()
 
-hpx_option(HPX_WITH_CXX11_AUTO_RETURN_VALUE BOOL
-  "Enable the use of auto as a return value in some places. Overriding this flag is only necessary if the C++ compiler is not standard compliant, e.g. nvcc."
-  ON ADVANCED)
-if(HPX_WITH_CXX11_AUTO_RETURN_VALUE)
-  hpx_add_config_define(HPX_HAVE_CXX11_AUTO_RETURN_VALUE)
-endif()
-
 ################################################################################
 # check for miscellaneous things
 ################################################################################

```

**Comment by hkaiser** [comment]
@gentryx: also, may I suggest to rewrite the patch in the following fashion:
```
diff --git a/hpx/parallel/executors/execution_fwd.hpp b/hpx/parallel/executors/execution_fwd.hpp
index 5df2085aa9..1fc00821e0 100644
--- a/hpx/parallel/executors/execution_fwd.hpp
+++ b/hpx/parallel/executors/execution_fwd.hpp
@@ -307,29 +307,29 @@ namespace hpx { namespace parallel { namespace execution
 
         ///////////////////////////////////////////////////////////////////////
         // async_execute customization point
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+#ifdef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+        template <typename Executor, typename F, typename ... Ts>
+        HPX_FORCEINLINE
+        auto customization_point<async_execute_tag>::operator()(
+            Executor && exec, F && f, Ts &&... ts) const
+        {
+            return async_execute(std::forward<Executor>(exec),
+                std::forward<F>(f), std::forward<Ts>(ts)...);
+        }
+#else
         template <>
         struct customization_point<async_execute_tag>
         {
         public:
-#endif
             template <typename Executor, typename F, typename ... Ts>
             HPX_FORCEINLINE
-            auto
-#ifdef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
-            customization_point<async_execute_tag>::
-#endif
-            operator()(
-                Executor && exec, F && f, Ts &&... ts) const
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
+            auto operator()(Executor && exec, F && f, Ts &&... ts) const
             -> decltype(async_execute(std::forward<Executor>(exec),
                     std::forward<F>(f), std::forward<Ts>(ts)...))
-#endif
             {
                 return async_execute(std::forward<Executor>(exec),
                     std::forward<F>(f), std::forward<Ts>(ts)...);
             }
-#ifndef HPX_HAVE_CXX11_AUTO_RETURN_VALUE
         };
 #endif
 
```

yes, this adds some duplication of code but significantly enhances readability.

**Comment by hkaiser** [comment]
Otherwise, this LGTM, thanks a lot for your persistence and patience!

**Comment by hkaiser** [comment]
>  I've now switched my HPX builds to use Ubuntu 12.04 if that's required to get this merged. You should file a separate ticket to make sure HPX builds with 14.04 too because I assume that 12.04 will some day no longer be available.

I don't think this is an OS version issue. It looks like the newer OS version does not come with the Perl module which is causing the problems. In the end it does not really matter as the docker image is maintained by us, so nobody will change the used OS without us knowing.

**Comment by gentryx** [comment]
@hkaiser 

- WRT setting CUDA_NVCC_FLAGS: yes, that would be one solution. It's got side effects though as you then assume that the user enables HPX' components for CUDA (-DHPX_WITH_CUDA=true), which may not be what a user wants. Example: If I set this option then HPX will export the flag "--default-stream per-thread" via CUDA_NVCC_FLAGS. This can break existing CUDA code in subtle ways. I'd say: this is opening another can of worms.

- WRT code duplication vs. readability: TBH since CUDA builds don't seem to be in your build rotation I specifically wanted to avoid having a header where someone could change one code path and leave the other unchanged.

- WRT OS versions: yes, it's more precise to say that it's not Ubuntu 14.04 that's causing the breakage here, but the image of Ubuntu 14.04 that Circle CI is using. The image is selected in the settings of each user, the docker image is more like an overlay, right? At least we didn't quite spot this setting in my builds weeks.

**Comment by hkaiser** [comment]
> WRT setting CUDA_NVCC_FLAGS: yes, that would be one solution. It's got side effects though as you then assume that the user enables HPX' components for CUDA (-DHPX_WITH_CUDA=true), which may not be what a user wants. Example: If I set this option then HPX will export the flag "--default-stream per-thread" via CUDA_NVCC_FLAGS. This can break existing CUDA code in subtle ways. I'd say: this is opening another can of worms.

Yes, you're right. Could we at least make this configuration setting `ON` whenever CUDA/NVCC is enabled, leaving it ``OFF` by default? Otherwise the user would have to specify two configuration options in order to build HPX with CUDA.

> WRT code duplication vs. readability: TBH since CUDA builds don't seem to be in your build rotation I specifically wanted to avoid having a header where someone could change one code path and leave the other unchanged.

FWIW, we have CUDA builds, but those are broken for weeks because of this issue .... ;-)

Adding a comment warning the user to change both code path when needed should take care of this. The way it is right now the code is close to impossible to grasp, at least for me.

**Comment by gentryx** [comment]
K, I've refactored the patch for readability as you suggested. The new CXX11 auto option is now off by default if HPX_WITH_CUDA is selected.

**Comment by hkaiser** [comment]
> K, I've refactored the patch for readability as you suggested. 

Thanks, perfect!

> The new CXX11 auto option is now off by default if HPX_WITH_CUDA is selected.

Uhh, shouldn't it be `ON` by default if CUDA is selected, but `OFF` by default otherwise?

**Comment by hkaiser** [comment]
@gentryx: uhh, I moved this over to the branch https://github.com/STEllAR-GROUP/hpx/tree/gentryx-master in the main repo and fixed some things. I'll try to reintegrate your work with that branch tomorrow. Thanks!

**Comment by gentryx** [comment]
@hkaiser Okidoki. I've just pushed some changes to please inspect. And yes: the option should be on by default and off if CUDA is being used as NVCC doesn't support it.

**Comment by gentryx** [comment]
@hkaiser Regarding your branch gentryx-master: the default value for the option needs to be inversed: on by default, off if CUDA (with gcc) is selected.

**Comment by hkaiser** [comment]
@gentry: ok, I'll verify again.

**Comment by hkaiser** [comment]
@gentryx I changed the default for the option (thanks for pointing this out) and reintegrated all of your recent changes into https://github.com/STEllAR-GROUP/hpx/tree/gentryx-master (please check, I hope I have not missed anything). Once circleci passes I'm going to merge things from there.

**Comment by gentryx** [comment]
@hkaiser Looks good so far. I've merged the gentryx-master branch in this PR and added another commit to mask some tests that won't compile in CXX11-mode (as they expect auto-typed lambda parameters).

**Comment by hkaiser** [comment]
@gentryx: thanks so much!

**Comment by gentryx** [comment]
I have no clue what the error message means that CircleCI is showing. Does this block the merge?

**Comment by hkaiser** [comment]
@gentry: I think this is happening because the build - for some reason - runs under your account, not under the STEllAR-GROUP account. I think we're fine merging anyways.

**Comment by gentryx** [comment]
Woohoo! :-)
